website/content/functional-database.md

```diff
@@ -0,0 +1,299 @@
+let my_var = 42;
+```
+
+In conventional programming languages[1], there is a difference between using
+`my_function()` and `my_var`. The first one evaluates the expression at the call
+site, while the second one evaluates it at the declaration site. It is also
+possible to express just `my_function`, a reference to a function that can be
+called without arguments.
+
+But if all your functions are pure, we generalize by saying that all three cases
+are equivalent.
+
+To make this work, let's say that `my_function` is "implicitly invoked" just as
+it would have been expressed as `my_function()`.
+
+Also, if we declare all functions to be pure it doesn't matter exactly _when_
+they are evaluated. So let's give the compiler the authority to make an informed
+guess about _when_ and make `my_var` and `my_function()` semantically
+equivalent.
+
+[1]: Let's say that conventional languages are first 15 from this list:
+https://survey.stackoverflow.co/2022/#most-popular-technologies-language-prof
+
+## Syntax, surly shinier
+
+First of all, lets change the function call to this:
+
+```
+(my_function arg1 arg2 arg3)
+```
+
+It may look strange because function name is within the parenthesis and there is
+no commas between arguments. But trust me, there are benefits to this syntax.
+
+Firstly, when the call has no arguments, the parenthesis can be omitted:
+
+```
+(my_function) == my_function
+```
+
+which feels very natural because of the similar behavior with expressions. As
+intended, function call with no argument and a plain reference to the function
+are expressed with the same syntax.
+
+Let's extend this behavior and allow allow bare function calls in a few places
+where they cannot become ambiguous:
+
+```
+# a list:
+[my_function arg1 arg2 arg3, my_function arg4 arg5 arg6]
+# a declaration:
+let res = my_function arg1 arg2 arg3
+```
+
+Secondly, I'd argue that currying looks natural.
+
+```
+let curry = my_function arg1 arg2
+let res = curry arg3
+
+# or inline:
+let res = (my_function arg1 arg2) arg3
+```
+
+Now let's go one step further and introduce a "pipe" operator. It applies left
+operand as an argument to the right operand:
+
+```
+arg3 | my_function arg1 arg2
+```
+
+This is especially useful for chaining function calls. If we declare `div`,
+`floor` and `mul` as common arithmetic functions, the function call syntax
+really starts to make sense:
+
+```
+12 | div 5 | floor | mul 5
+```
+
+## Running real relations
+
+Up to this point, we were talking language design without a clear justification.
+Language being cool is fine, but it is much more important that the language is
+actually the right tool for the job.
+
+The job, in the case of PRQL, is querying databases.
+
+Core unit of data here is a relation: an ordered set of rows, each of which
+contains entries for each of the columns.
+
+Queries operate in a static environment that contains references to database
+tables and a library of standard functions.
+
+```
+mod default_db {
```

Comment:
I find this a bit confusing, but maybe it's just me.

I think the article is strong because it has the potential to persuade data people that:
a) functional languages are good
b) functional languages can be applied to relational data transformations
c) functional languages can work well with SQL, by operating at the relational level — there's no need to write a whole Scala framework with its own data processing engine...

That's already a lot! So anything else, like `default_db` we could try and minimize? I don't think folks will be confused by where `albums` come from?

Reply:
I'm inclined to agree on the default_db point. It was very interesting for me personally to see how the `from tbl` translates into `default_db.tbl` but it does seem a bit orthogonal to the rest of the argument (from what I recall right now, would have to reread to be sure).

Reply:
Maybe you are right - it's not the main point of the article and but code block attracts too much attention.